File: benchmarks/app/src/model/quick_sort.cpp
#include <array>
#include <random>

#include "quick_sort.hpp"

namespace model {

    namespace {

        uint64_t find_pivot(std::vector<uint64_t>& data, int leftBorder, int rightBorder) {
            static std::random_device seeder;
            static std::mt19937 rng(seeder()); 
            std::uniform_int_distribution<int> gen(leftBorder, rightBorder - 1);
            const int idx = gen(rng);

            std::swap(data[idx], data[rightBorder - 1]);

            return data[rightBorder - 1];
        }

        int place_pivot(std::vector<uint64_t>& data, int leftBorder, int rightBorder) {
            int index = leftBorder - 1;
            const uint64_t pivot = find_pivot(data, leftBorder, rightBorder);

            for (int x = leftBorder; x < rightBorder; ++x) {
                if (data[x] <= pivot) {
                    index += 1;
                    std::swap(data[x], data[index]);
                }
            }

            return index;
        }

        void sort(std::vector<uint64_t>& data, int leftBorder, int rightBorder) {
            if (leftBorder >= rightBorder || data.size() < 2) {
                return;
            }

            const int pivot = place_pivot(data, leftBorder, rightBorder);

            sort(data, leftBorder, pivot);
            sort(data, pivot + 1, rightBorder);
        }

    }

    void quick_sort(std::vector<uint64_t>& data) {
        sort(data, 0, data.size());
    }

}

File: presentation/src/model/sorting/quicksort.ts
export default function quicksort<T>(data: T[], comparator: (a: T, b: T) => boolean) {
    let u = 0
    const sort = (arr: T[], leftBorder: number, rightBorder: number) => {
        console.log(leftBorder, rightBorder)
        if (u++ > 10) {
            return
        }
        if (leftBorder >= rightBorder || arr.length < 2) {
            return
        }

        const pivotIndex = placePivot(arr, leftBorder, rightBorder)
        sort(arr, leftBorder, pivotIndex)
        sort(arr, pivotIndex + 1, rightBorder)
    }

    const placePivot = (arr: T[], leftBorder: number, rightBorder: number) => {
        let index = leftBorder - 1
        let pivot = findPivot(arr, leftBorder, rightBorder)

        for (let x = leftBorder; x < rightBorder; ++x) {
            if (!comparator(arr[x], pivot)) {
                index += 1
                ;[arr[x], arr[index]] = [arr[index], arr[x]]
            }
        }

        return index
    }

    const findPivot = (arr: T[], leftBorder: number, rightBorder: number) => {
        let high = rightBorder - 1
        let rng = Math.floor(Math.random() * (rightBorder - leftBorder) + leftBorder)

        ;[arr[rng], arr[high]] = [arr[high], arr[rng]]

        return arr[high]
    }

    sort(data, 0, data.length)
    return data
}
